include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,426 @@
+    using _FlagStorageType = typename _FlagType::_FlagStorageType;
+
+    using _KernelName = typename _KernelParam::kernel_name;
+    using _LookbackInitKernel =
+        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<__lookback_init_kernel<_KernelName>>;
+    using _LookbackKernel =
+        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<__lookback_kernel<_KernelName>>;
+
+    const ::std::size_t __n = __in_rng.size();
+
+    if (__n == 0)
+        return sycl::event{};
+
+    static_assert(_Inclusive, "Single-pass scan only available for inclusive scan");
+    static_assert(oneapi::dpl::unseq_backend::__has_known_identity<_BinaryOp, _Type>::value,
+                  "Only binary operators with known identity values are supported");
+
+    assert("This device does not support 64-bit atomics" &&
+           (sizeof(_Type) < 64 || __queue.get_device().has(sycl::aspect::atomic64)));
+
+    // Next power of 2 greater than or equal to __n
+    auto __n_uniform = ::oneapi::dpl::__internal::__dpl_bit_ceil(__n);
+
+    // Perform a single-work group scan if the input is small
+    if (oneapi::dpl::__par_backend_hetero::__group_scan_fits_in_slm<_Type>(__queue, __n, __n_uniform))
+    {
+        return oneapi::dpl::__par_backend_hetero::__parallel_transform_scan_single_group(
+            oneapi::dpl::__internal::__device_backend_tag{},
+            oneapi::dpl::execution::__dpl::make_device_policy<typename _KernelParam::kernel_name>(__queue),
+            ::std::forward<_InRange>(__in_rng), ::std::forward<_OutRange>(__out_rng), __n,
+            oneapi::dpl::__internal::__no_op{}, unseq_backend::__no_init_value<_Type>{}, __binary_op,
+            ::std::true_type{});
+    }
+
+    constexpr ::std::size_t __workgroup_size = _KernelParam::workgroup_size;
+    constexpr ::std::size_t __data_per_workitem = _KernelParam::data_per_workitem;
+
+    // Avoid non_uniform n by padding up to a multiple of workgroup_size
+    ::std::size_t __elems_in_tile = __workgroup_size * __data_per_workitem;
+    ::std::size_t __num_wgs = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __elems_in_tile);
+
+    constexpr int __status_flag_padding = SUBGROUP_SIZE;
+    ::std::size_t __status_flags_size = __num_wgs + 1 + __status_flag_padding;
+
+    ::std::size_t __status_flags_size_mem_size =
+        __status_flags_size * sizeof(_FlagStorageType) + (__status_flags_size * sizeof(_Type)) * 2;
+    ::std::size_t __status_vals_full_offset = __status_flags_size * sizeof(_FlagStorageType);
+    ::std::size_t __status_vals_partial_offset = __status_vals_full_offset + __status_flags_size * sizeof(_Type);
+
+    ::std::byte* __device_mem = sycl::malloc_device<::std::byte>(__status_flags_size_mem_size, __queue);
+    assert(__device_mem);
+
+    _FlagStorageType* __status_flags = reinterpret_cast<_FlagStorageType*>(__device_mem);
+    _Type* __status_vals_full = reinterpret_cast<_Type*>(__device_mem + __status_vals_full_offset);
```

Comment:
It seems that the pointer may be not aligned for `_Type` (i.e. its address may not be multiple to `_Type`), especially when `sizeof(_Type) > sizeof(_FlagStorageType)`, because the offset is `__status_flags_size * sizeof(_FlagStorageType)`.

Could you improve memory management to guarantee that the pointers are aligned for their types?

Reply:
Here is an example on [godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAMzSrgAyeAyYAHI%2BAEaYxCAAnKQADqgKhE4MHt6%2BAdKp6Y4CoeFRLLHxSXaYDplCBEzEBNk%2BfoG2mPZFDPWNBCWRMXGJtg1NLbntCmP9YYPlwwkAlLaoXsTI7Bzm/mHI3lgA1Cb%2Bbk7TxJisJ9gmGgCCO3sHmMenbCwkAJ43d49muwY%2By8RxObgIX2SmAA%2BgRiExCApfg8nkCXm83MhpvhBMj/oDgaDTqIFEomnjUYTXmCIVDYfDERSAc8QdTidisFQKSjmWjWRisQR0LQ8NEbocAPQSw7YkAgFhiWioUwoh4ETAsZIGdVg/ZMUmHAAqeOiqE8hzwCmhYjwwHC6AgaAY00OADdUHh0IcAFSHZJwpaHBioTCqTb%2B44AdisD0OccOTC8RAt/uIbwAIodLmF1cRkpcCNCSQQwbKQF4c6nYTcIKmlicY/d41nMAR1gxDmBIHhU8czABWBMiu2oKgQQ1Lev%2BRsmSPpv4LgGcuZG7BCQ3Q%2B7BACSAHEIth56r7jnDgqwhB6yjo39m1LDthVKwta8wskkwoAHTf2/xrzpIxDmhQ1ITZTN0DWaJ6AbX843/MJgCAgAxAxgHqEgmGATAQKhDMZSFOUK0EfwzGradYPw9A5XSAAvGECCA6FpiYNsrSoVCrVosDDkkSMYOPRddioQ5l3tcc1w3AB5AAlPdtwiLcrybeNqIIkAuNhRjmNY6F2Mwzi8Do6EPiYwywIo5toSYhodL04ADLon0ZTM0cIGhFDMPQ%2BEsJwzBA2sQ43Oslj/10jjTMc30uNc4DQMnJyzAbSVpUtQ4CAQV4DBdNBNQERgGP4NMPm%2BABaS5tUwL1kiYdAcWAE4kIo1SqPUszNKs7TQtdMQ2K8WhaGhUcqCUBiTkzDqbNCuyHNeKKXLHdzUK8zDsLi/jlLjZrqLahiJpCq1utoK1qqaPAxEGqhhtbPC9p0w7ev6i6rtGywtMmtjwq4pzooW3ypxnWMVPUtToi%2BdVfSsrBXTwTZjI1PDs0EOJ81bIt9RLU4tpAUHwZrMsFX65Ugs6j79IimETK4yd1ubfUyQIIKoZhimNX%2BhdAbjRbPKIbzVqhCHgtsjiEcwHNkYLNHpjBLm0J5lbfO9GtIcwaHYY%2BNmOaAhW3v26F7t0vraBFsW8wl4tpYVpXoSZtX4YC26up6g3HqGkaNY2rXQIFkm9adk7HDEY2kdN1HzdOWL%2Batm2WZYPsLB1u6/b6M6Btd1t3Z2Vw8C5ATmWE0TKvE9dNx3fdDyU5ssY0hiTJtO1oWqr0xucuiYr%2BmmgbLautN16boRxzAFBuwWps%2BszvvmtyPNljCfLW8jNarnaE8do7ndTy6Rv7sHB%2BHn2%2B6%2BubW9%2B%2BfG0r4GWu7h2DqT07zrTwsB6H5vr7CsnD5bzA29PpqL%2B2ozdpwxYNvdUz9/DjSYr3cKT844rxvmvKghsnpbxgfbEe8DjrJ3vpvVGqDXq12HAwBuNV1oUXvMrVWMcLRDzEJcGqXwhy2ntIcQqCYGAMIUKIfqjQ0qgVIEGVACZaoZAEIHaYmBkg0MuEGTAlVKosJIGlDKLC8DEBdEoAAjl4Rgmxf5lgHgLaOQCg65hRoWMObgsYGLxmpAmSpkBBRMk/amC8PZ0ziAzChzMgEZ01jLZac9%2BZwLfvZEx4tQ7o2ltPAJfNMCK38NgRmKtvHqw7ptP%2BrUAGMUuOeBgWA0wvyASA3eJUe5CzJk/NJboPToAFtEfUMJezN0RqYs2kTTjuk9PExJXjbaxzQQoKBFSd4KHds2TptTGJ13tA3OEeEsbTIgAswhbBBAXT1hbUCNx%2BGv31ogl2ODH4jJ2f3BpsyyDZI1AiPJcQxnxgjnE4JezDZhJDuY9pbgHndKCtMyq5y7mc21rs2%2BAcjbNNFsHMxksMafMtgk4m71fYIJeQMxF%2Bt/Yp2QbgkZGcARZxzo8TW94xokvAaSzMbgMrIAANYZnJSSzWZY0BJgxGCPspF0EhIUCAdlrLTjBOmscTWzYRV8rcOy/hlprSEMqjy8wZgxWBSlb89AUSlpy0CZgK2%2B8OLxTBGyssrhaBVKZWsUapw2XyuBci/qcqzAKv1fy61D0wXCtFfGR14r5WSqtCqu1DqLX8ogMqmVqrw6%2BW1Wip2%2BzaB6sDeKw1eTjWuObKallnr2XOuIXfWg/rFVZoxYHCy7q4wZu9dQ6VTDZW8ozcG31obNlQkjbrdFWDY3%2BXjZROURqTVqSoJcTASTKE%2BNIXODgKxc0cH7LwPwHAtCkFQJwNw1gAoKDWBsakAIeCkAIJocdKxqUgEkAANk/JIfw/h%2BzHv7JGDQx7/AaEjAkSM0hJ2SBnXuhdnBeDco0DuvdKw4CwBgIgEAoYahJhIOQSgjR7LKEMJ0IQCBUAAHdZ3bpyskOgLFMjwfCLQJDqHZ3zsw3QYYwAuAkVIKR%2BgxAIisC2LwGjcRJJJkI2hzg27wPIHuMQey37SDcfqDiAT/BBAiDEOwKQMhBCKBUOoOdOg9AGCMCgFdlh9Cim5ZAFYqB/SZG5RwXgqBXRxGIJ6Qe8AVjEC8IIPAbBDRmiOlZ1Y6xNh6GxGEPDiHkMce4LwFD8JkicfHfoTg07SDEeM5wR8EGeaHFUAADmPSVY9khDjAGQMgQ4lHPwKogMuyw1h%2BG4EIIonYXAli8F3YpycpBD1mES5%2BK9kYBwPqfSRfwx771hY4O%2B0gLAQD9j/VFr9RnbAgD/TVrQgGQMQCQGa98BBoOOlQJqMjdGGOcCSyltLGWss5f8Hl3glUyvmaolwGTwguFSekGJ%2BQSg1Cft0JdwLTBgv%2BYneFj9imxusYIEtw4o4EvJdS%2BlzL2Xcv5Y8Ot2jfZ/CVeqwBlYGUarDEvL1/rg3huRc/Yu8bv7/21YPSAMwkYz2Ri4AkRLJ7b2JcaxoBI/hev%2BB%2B/O/HSPaukCA6B5jZAKCrdh%2BRyjZg%2BB0FzNyiA0RP3RDCI0L4IXSCy%2BYMQL4klojaBqDVjDa3VkEEkgwWgCvftYGiF4YAbhFSGe3VgBURhxAm9UVrvApnDPzu40mLY26xaTvnSKaI8JVceCwJ%2BuE9nFemeIKaJQ6Yrn24QgBvgHEABqeBMAocklCdDvB7sSfENJ%2B7cmnu/Ze/oQwxh1M2H99pjHenuiGeM5H87lmdMdC6JkFweSJh%2BEuyEOYZQKh6AKCIrInhWhD7SCPgYA/hiXeqLUAQvRxhj9yHPzozvF8zGn0MeIc%2BZjd481v/vO%2BJArDXW5qTX2p1s%2BixwEHu3wcHah4FUr8WKtVaJzNknZhT2M8kP/hIY9e1ZLfwJnV9TgfrUbDnCbKbZHXrMwG/MbTnL/UgSPdIZwSQIAA). I can make a patch.

Reply:
Thanks for the godbolt example. I can incorporate what you have written into this branch.

Reply:
Here is the patch: [aligned.patch](https://github.com/oneapi-src/oneDPL/files/15014668/aligned.patch)

Reply:
Thanks for the patch. I applied it.